src/bytes.rs

```diff
@@ -200,6 +201,53 @@ impl Bytes {
         }
     }
 
+    /// Create [Bytes] with a buffer whose lifetime is controlled
+    /// via an explicit owner.
+    ///
+    /// A common use case is to zero-copy construct from mapped memory.
+    ///
+    /// ```ignore
+    /// use bytes::Bytes;
+    /// use std::fs::File;
+    /// use memmap2::Map;
+    ///
+    /// let file = File::open("upload_bundle.tar.gz")?;
+    /// let mmap = unsafe { Mmap::map(&file) }?;
+    /// let b = Bytes::from_owner(mmap);
+    /// ```
+    ///
+    /// The `owner` will be transferred to the constructed [Bytes] object, which
+    /// will ensure it is dropped once all remaining clones of the constructed
+    /// object are dropped. The owner will then be responsible for dropping the
+    /// specified region of memory as part of its [Drop] implementation.
+    ///
+    /// Note that converting [Bytes] constructed from an owner into a [BytesMut]
+    /// will always create a deep copy of the buffer into newly allocated memory.
+    pub fn from_owner<T: AsRef<[u8]>>(owner: T) -> Self {
```

Comment:
Doesn't this need some kind of StableDeref bound to protect against weird edge cases where the referenced region changes over time?

Reply:
The `string` crate had a similar issue: https://github.com/carllerche/string/issues/9

I do wonder if `Sync` would be sufficient here... In general, we probably do need a `Send` bound as well since `Bytes` is `Send`

Reply:
Sync wouldn't cover it since you can have a thread-safe implementation that still mutates internally.

Reply:
Hmm yeah only calling as_ref once may actually make this fine.

Reply:
StableDeref is *not* necessary because we don't call `as_ref` more than once.

Reply:
There's a safe way for it to still be UB, even with only calling it once. If I pass in a form of `Arc<Mutex<_>>`, and then afterwards change the internal buffer completely (maybe a reallocation), the captured pointer could be referring to invalid memory.

Reply:
The reference we get back from the as_ref needs to be valid for the remaining lifetime of the object though - it can't be swapped out without a call to an &mut method that would ensure the initial borrow is gone.

Reply:
No, it's sound as-is. You cannot write a safe thing where this breaks even with `Arc`/`Mutex` because if the buffer is inside the mutex, the implementer of `AsRef` can't return the buffer from `as_ref` since it borrows from the mutex guard.

As written, the `as_ref` call essentially borrows the `owner` field until we run the destructor. As long as we never make any mutable function calls on `owner` during that time, it's okay.

Reply:
```rs
use std::sync::{Arc, Mutex};

struct Evil {
    within: Arc<Mutex<Vec<u8>>>,
}

impl AsRef<[u8]> for Evil {
    fn as_ref(&self) -> &[u8] {
        // error[E0515]: cannot return value referencing temporary value
        self.within.try_lock().unwrap().as_ref()
    }
}
```

Reply:
If I'm not mistaken the current API should guard against `Owner` exposing a thread local as well.